File: source/yeon/4주차/code/radixSort.cpp
#include <iostream>
#include <string>
#include <random>
#include <algorithm>
#define LEN 8 
using namespace std;

void countingSort(int arr[], int digit){
    int sorted[LEN];
    int count[10] = {0,};

    for(int i = 0; i < LEN; i++){
        int index = arr[i]/digit;
        count[index%10] += 1;
    }

    for(int i = 1; i < 10 ; i ++){
        count[i] += count[i-1];
    }
    
    int a = LEN -1;
    while(a >= 0){
        int index = arr[a] / digit;
        sorted[count[index % 10] - 1] = arr[a];
        count[index % 10] -= 1;
        a--;
    }
    for(int i = 0; i < LEN ; i++){
        arr[i] = sorted[i];
    }
}

void radixSort(int arr[], int len){
    int digit = 1;
    int max_value = *max_element(arr,arr+len);

    while(max_value/digit > 0){
        countingSort(arr,digit);
        digit *= 10;
    }
}

int main(void){
    srand(time(NULL));
    int arr[LEN];
    printf("정렬 전 : ");
    for(int i = 0 ; i < LEN; i++){
        arr[i] = rand()%100+1;
        printf("%d ", arr[i]);
    }
    printf("-> 정렬 후 : ");
    radixSort(arr, LEN);
    for(int i = 0; i< LEN; i++)
        printf("%d ", arr[i]);

    return 0;
}

File: source/yeon/4주차/code/radixSort.py
def countingSort(arr, digit):
    n = len(arr)

    #arr that will contain sorted arr
    sorted = [0]*n
    #arr that count the number of how many elements are in digit
    count = [0]*10

    #counting how many elements are in each digit 
    for i in range(0,n):
        index = int(arr[i] / digit)
        count[index % 10] += 1
    
    #cumulative count element at count arr
    for i in range(1,10):
        count[i] += count[i-1]
    
    i = n -1
    while i >= 0:
        index = int(arr[i]/digit)
        sorted[count[index%10]-1] = arr[i]
        count[index%10] -= 1
        i -= 1

    for i in range(0,len(arr)):
        arr[i] = sorted[i]

def radixSort(arr):
    #find the max value at arr
    max_v = max(arr)
    #digit is value, dividing the max value with multiply 10 each cycle
    digit = 1
    # repeat until the max value's quotient is bigger than 0 
    while int(max_v/digit) > 0 :
        countingSort(arr,digit)
        digit *= 10

from random import randint
arr = [randint(1,101) for i in range(8)] # 1부터 101사이의 랜덤 값을 8개의 list 안에 초기화
print("정렬 전 : ", arr, end=" ") 
radixSort(arr)
print ("-> 정렬 후 : ",arr) 
